crates/matrix-sdk/src/room/timeline/event_handler.rs

```diff
@@ -754,10 +756,71 @@ impl NewEventTimelineItem {
 
     fn room_member(
         user_id: OwnedUserId,
-        content: FullStateEventContent<RoomMemberEventContent>,
+        full_content: FullStateEventContent<RoomMemberEventContent>,
         sender: OwnedUserId,
     ) -> Self {
-        Self::from_content(TimelineItemContent::RoomMember(RoomMember { user_id, content, sender }))
+        use ruma::events::room::member::MembershipChange as MChange;
+        let item_content = match &full_content {
+            FullStateEventContent::Original { content, prev_content } => {
+                let membership_change = content.membership_change(
+                    prev_content.as_ref().map(|c| c.details()),
+                    &sender,
+                    &user_id,
+                );
+
+                if let MChange::ProfileChanged { displayname_change, avatar_url_change } =
+                    membership_change
+                {
+                    TimelineItemContent::ProfileChange(MemberProfileChange {
+                        user_id,
+                        displayname_change: displayname_change.map(|c| Change {
+                            new: c.new.map(ToOwned::to_owned),
+                            old: c.old.map(ToOwned::to_owned),
+                        }),
+                        avatar_url_change: avatar_url_change.map(|c| Change {
+                            new: c.new.map(ToOwned::to_owned),
+                            old: c.old.map(ToOwned::to_owned),
+                        }),
+                    })
+                } else {
+                    let change = match membership_change {
+                        MChange::None => MembershipChange::None,
+                        MChange::Error => MembershipChange::Error,
+                        MChange::Joined => MembershipChange::Joined,
+                        MChange::Left => MembershipChange::Left,
+                        MChange::Banned => MembershipChange::Banned,
+                        MChange::Unbanned => MembershipChange::Unbanned,
+                        MChange::Kicked => MembershipChange::Kicked,
+                        MChange::Invited => MembershipChange::Invited,
+                        MChange::KickedAndBanned => MembershipChange::KickedAndBanned,
+                        MChange::InvitationAccepted => MembershipChange::InvitationAccepted,
+                        MChange::InvitationRejected => MembershipChange::InvitationRejected,
+                        MChange::InvitationRevoked => MembershipChange::InvitationRevoked,
+                        MChange::Knocked => MembershipChange::Knocked,
+                        MChange::KnockAccepted => MembershipChange::KnockAccepted,
+                        MChange::KnockRetracted => MembershipChange::KnockRetracted,
+                        MChange::KnockDenied => MembershipChange::KnockDenied,
+                        MChange::ProfileChanged { .. } => unreachable!(),
+                        _ => MembershipChange::NotImplemented,
```

Comment:
I think it could deserve a `From` impl to translate from `MChange` to `MemberChange`, what do you think?

I understand there is a special case here for `ProfileChanged`, but since it's unreachable, it could just simplify the code to write `change: Some(change.into())`.

Reply:
We could but implementing `From` makes it part of the public API, so it would not be unreachable anymore.

However `ProfileChanged` in that case would translate to `MembershipChange::None` (since a profile change means the membership didn't change) so it could be done.

Alternatively, if we really want that code to be split somewhere else but don't want the conversion as part of the public API we could declare a private function on `MembershipChange`.